types/shallowequal/index.d.ts

```diff
@@ -2,14 +2,24 @@
 // Project: https://github.com/dashed/shallowequal
 // Definitions by: Sean Kelley <https://github.com/seansfkelley>
 //                 BendingBender <https://github.com/BendingBender>
+//                 Arnd Issler <https://github.com/arndissler>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.3
 
 declare function shallowEqual<TCtx = any>(
     objA: any,
     objB: any,
-    compare?: (this: TCtx, objA: any, objB: any, indexOrKey?: number | string) => boolean | void,
+    compare?: shallowEqual.Comparer<TCtx>,
     compareContext?: TCtx
 ): boolean;
 
+declare namespace shallowEqual {
+    type Comparer<T = any> = (
+        this: T,
+        objA: any,
+        objB: any,
+        indexOrKey?: number | string
+    ) => boolean | undefined;
+}
```

Comment:
A namespace for this is a little heavyweight; what about just doing an `export type` at the top level?

Reply:
I'd like to go with the namespace approach, because it's not breaking existing codebases. Currently you can use multiple ways of importing the `shallowequal`, e.g.

`import shallowequal from 'shallowequal`
`import eq from 'shallowequal'`

If you export the type on top level, you break this, because you have to use

`import { shallowEqual } from 'shallowequal'`

or

`import { shallowEqual, Comparer } from 'shallowequal'`

So I think it's a good tradeoff to have a namespace which might not be necessary instead of breaking existing codebases.

Reply:
To me, it's kinda more idiomatic to not return anything instead of an `undefined` value. Why did you change this, was there a technical reason or purely aethetical?

Reply:
It's legal to mix default and non-default imports, so you should be able to `import eq, { Comparer } from 'shallowequal';` meaning it'll be backwards-compatible.

In any case, I don't mind either way aesthetically speaking, but I do know that namespace-merging is sometimes considered an anti-pattern, which is why I raised it.

I'm okay with these changes as-is; I'll defer to @BendingBender to approve since they've got other comments.

Reply:
The change I made is not aesthetical.

The shallowequal documentation clearly stated, that the `customizer` function shall return `true`, `false` or `undefined`. This is also covered by the comparison in the code of shallowequal: 
```javascript
if (ret === false || (ret === void 0 && valueA !== valueB)) {
  return false;
}
```

`void 0` is equal to `undefined`, while `void` may contain `null` or `undefined`. When using `boolean | void` as result for the compare function, implicit return of `undefined` is okayish. But it's also possible to return `null`, which is evaluated falsy. But based on the shallowequal code it's `true` for the comparison of `valueA` and `valueB`.

Since you may not have compared the values in your code but accidentally return `null` when `strictNullChecks` is `false` (which I assume is in many legacy codebases), it's a source of hard to find errors. I agree that the accidentally return of `null` is not very likely when `strictNullChecks`

But considering all this: that's why I changed the signature to `undefined`.

Reply:
Sorry, maybe I misunderstand you, but I've tested what you said and it seems not to be correct.

Following code was run with `"strictNullChecks": "true"`:
```typescript
const u: undefined = null; // error
const n: null = undefined; // error
const v: void = null; // error
const v1: void = undefined; // ok
const v2: void = void 0; // ok

function a(): void {
  return undefined; // ok
}

function b(): undefined {} // error

function c(): undefined {
    return null; // error
}
```

The same code with `"strictNullChecks": "false"`:
```typescript
const u: undefined = null; // ok
const n: null = undefined; // ok
const v: void = null; // ok
const v1: void = undefined; // ok
const v2: void = void 0; // ok

function a(): void {
  return undefined; // ok
}

function b(): undefined {} // error

function c(): undefined {
    return null; // ok
}
```

So actually, in non-`strictNullChecks` mode the only difference between `void` and `undefined` is that the former doesn't require a return value, while the latter requires an explicit return of `undefined`.

It doesn't protect you from incorrectly returning `null` in non-`strictNullChecks` mode.

Reply:
> I'd like to go with the namespace approach, because it's not breaking existing codebases. Currently you can use multiple ways of importing the `shallowequal`, e.g.
> 
> `import shallowequal from 'shallowequal`
> `import eq from 'shallowequal'`
> 
> If you export the type on top level, you break this, because you have to use
> 
> `import { shallowEqual } from 'shallowequal'`
> 
> or
> 
> `import { shallowEqual, Comparer } from 'shallowequal'`
> 
> So I think it's a good tradeoff to have a namespace which might not be necessary instead of breaking existing codebases.

@seansfkelley Actually, the way @arndissler implemented it is the only valid solution. You can't use a default export here because of what is described [here](https://github.com/DefinitelyTyped/DefinitelyTyped#a-package-uses-export--but-i-prefer-to-use-default-imports-can-i-change-export--to-export-default). And you can't have multiple `export` statements in a `d.ts` file that has a `export = ...` export.

Reply:
I totally agree that it's comfortable to not explicitly write a `return undefined`. 

According to how it is implemented in `shallowequal`, the only valid return values for the `customizer` function are `true`, `false`, `undefined`. If you **don't** enable `strictNullChecks` then the return types are `true`, `false`, `null` and `undefined`, which matches to `void`. My assumption was, that TypeScript's `strictNullCheck` will remove the nullable option from `boolean`, so I wrote `boolean | undefined` to avoid the nullability of `void`. 

My mistake was that I didn't expect TypeScript to remove `null` from `void` when using `strictNullChecks`. So I'll go with changing it back to `void`. Thanks for the discussion 👍 

Since I will make some changes to this PR, should I change the name of the parameter `compare` and it's type `Comparer` to `customizer` and `Customizer`, so it matches with shallowequals documentation?

Reply:
> Since I will make some changes to this PR, should I change the name of the parameter `compare` and it's type `Comparer` to `customizer` and `Customizer`, so it matches with shallowequals documentation?

Sounds reasonable to me.

Reply:
> My mistake was that I didn't expect TypeScript to remove `null` from `void` when using `strictNullChecks`. So I'll go with changing it back to `void`. Thanks for the discussion 👍

Thank you for being patient with me being pedantic.